### PR TITLE
Initial commit of the cmake extractor command

### PR DESCRIPTION
**Pull request by shahms**
This definitely needs some tests and refactoring, but I wanted to get some comments on how it stands now.

### CODE REVIEW

**Comment by danielmoy-google** [comment]
Also do you need to append #2861 in the description at all, or no?

File: kythe/go/extractors/config/runextractor/cmakecmd/cmakecmd.go [hunk 1]
```diff
@@ -14,43 +14,188 @@
...
-// flags for cmake extraction.
 func (c *cmakeCommand) SetFlags(fs *flag.FlagSet) {
-	// fs.StringVar(&c.someFlag, "some_flag", "default-value", "Flag Description")
+	fs.StringVar(&c.extractor, "extractor", "", "A required path to the extractor binary to use.")
+	fs.StringVar(&c.sourceDir, "sourcedir", ".", "A required path to the repository root. Defaults to the current directory.")
+	fs.StringVar(&c.buildDir, "builddir", "", "An optional path to the directory in which to build. If empty, defaults to a unique subdirectory of sourcedir.")
+}
+
+func (c *cmakeCommand) verifyFlags() error {
+	for _, key := range []string{"KYTHE_CORPUS", "KYTHE_ROOT_DIRECTORY", "KYTHE_OUTPUT_DIRECTORY"} {
```
[the remark this answers is not in the document]
Comment (shahms) [reply kythe/go/extractors/config/runextractor/cmakecmd/cmakecmd.go]: I also want to check for the existence of KYTHE_OUTPUT_DIRECTORY, but don't currently (the extractor fails with a stack trace if it doesn't exist, which isn't terribly usefule).

File: kythe/go/extractors/config/runextractor/cmakecmd/cmakecmd.go [hunk 2]
```diff
@@ -14,43 +14,188 @@
...
+		if err != nil {
+			return c.Fail("unable to parse command line: %v", err)
+		}
+		cmd.Args = append(cmd.Args, args...)
+		cmd.Dir, err = filepath.Abs(entry.Directory)
+		if err != nil {
+			return c.Fail("unable to resolve cmake directory: %v", err)
+		}
+		cmd.Env = exEnv
+		// TODO(shahms): Parallelize this.
```
[the remark this answers is not in the document]
Comment (shahms) [reply kythe/go/extractors/config/runextractor/cmakecmd/cmakecmd.go]: I just implemented it instead (but should probably pull it out to a separate function).

**Comment by shahms** [comment]
Addresses #2861 (as long as it's mentioned somewhere, they should be linked).

File: kythe/go/extractors/config/runextractor/cmakecmd/cmakecmd.go [hunk 3]
```diff
@@ -14,43 +14,203 @@
...
+	commands, err := readCommands(filepath.Join(buildDir, "compile_commands.json"))
+	if err != nil {
+		return c.Fail("unable to read compile_commands.json: %v", err)
+	}
+	exEnv, err := extractorEnv(buildDir)
+	if err != nil {
+		return c.Fail("unable to set up environment: %v", err)
+	}
+	sem := semaphore.NewWeighted(128) // Limit concurrency.
+	extraction, ctx := errgroup.WithContext(ctx)
```
Comment (creachadair) [inline kythe/go/extractors/config/runextractor/cmakecmd/cmakecmd.go]: I suggest pulling this out into a function, though of course it works fine as-is.

File: kythe/go/extractors/config/runextractor/cmakecmd/cmakecmd.go [hunk 4]
```diff
@@ -14,43 +14,203 @@
...
 
-	// // Anything set from flags goes here
-	// someFlag string
+	extractor string
+	buildDir  string
+	sourceDir string
+}
+
+type compileCommand struct {
+	Command   string
```
[the remark this answers is not in the document]
Comment (shahms) [reply kythe/go/extractors/config/runextractor/cmakecmd/cmakecmd.go]: They cannae as they need to be exported (apparently) for json.Unmarshal to do a thing.

File: kythe/go/extractors/config/runextractor/cmakecmd/cmakecmd.go [hunk 5]
```diff
@@ -14,43 +14,203 @@
...
+	}
+	return commands, nil
+
+}
+
+// cmakeEnvironment copies the existing environment and modifies it to be suitable for an extractor invocation.
+func extractorEnv(buildDir string) ([]string, error) {
+	var env []string
+	for _, value := range os.Environ() {
+		// Until kzip support comes along, we only support writing to a single directory so strip these options.
```
[the remark this answers is not in the document]
Comment (shahms) [reply kythe/go/extractors/config/runextractor/cmakecmd/cmakecmd.go]: Done.

File: kythe/go/extractors/config/runextractor/cmakecmd/cmakecmd.go (hunk 3 above)
Comment (shahms) [reply kythe/go/extractors/config/runextractor/cmakecmd/cmakecmd.go]: Was in the middle of that when you responded; done.

**Comment by shahms** [comment]
PTAL

File: kythe/go/extractors/config/runextractor/cmakecmd/cmakecmd.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (creachadair) [reply kythe/go/extractors/config/runextractor/cmakecmd/cmakecmd.go]: Oh, yeah. I'd probably document that, either with a comment or by adding tags to the fields, e.g.,

```
// A compileCommand holds the decoded arguments of a CMake JSON command spec.
type compileCommand struct {
    Command string `json:"command"`
    Directory string `json:"directory"`
}
```

obviously not both are necessary, I'm just illustratin'.

[the remark this answers is not in the document]
Comment (shahms) [reply kythe/go/extractors/config/runextractor/cmakecmd/cmakecmd.go]: Done.

**Pull request merged**